### PR TITLE
Pickle redis

### PR DESCRIPTION
**Pull request by ownzandy**

### CODE REVIEW

File: redis_store.py [hunk 1]
```diff
@@ -1,11 +1,31 @@
...
+		session = pickle.dumps(session)
+		self.redis.execute_command('MULTI')
+		self.redis.set(session_name, session)
+		result = self.redis.execute_command('EXEC')
+		self.redis.execute_command('WATCH ' + session_name)
+		return result
+
+	def get_curr_session(self, session_name):
+		curr_session = self.redis.get(session_name)
+		self.redis.execute_command('WATCH ' + session_name)
```
Comment (1heart) [inline redis_store.py:25]: SQL injections are possible

File: hello.py [hunk 2]
```diff
@@ -34,50 +31,59 @@ def init_judge_session():
...
-        result['error'] = 'Need to init first!'
-        return jsonify(result)
     if 'judge_id' not in json_args or 'favored' not in json_args:
         result['error'] = 'Not enough args'
         return jsonify(result)
     judge_id, favored = json_args['judge_id'], json_args['favored']
-    result['error'] = curr_session.perform_decision(judge_id, favored)
+    result['error'] = g.curr_session.perform_decision(judge_id, favored)
+    if store.save_session(g.curr_session, g.session_name) is None:
+        before_request()
```
Comment (1heart) [inline hello.py]: Potential infinite loop if before_request doesn't resolve `store.save_session(g.curr_session, g.session_name)` being 'None'

Depends on your level of optimism

File: hello.py [hunk 3]
```diff
@@ -34,50 +31,59 @@ def init_judge_session():
...
         result['error'] = 'Not enough args'
         return jsonify(result)
     judge_id = json_args['judge_id']
-    a,b = curr_session.get_decision(judge_id)
+    a,b = g.curr_session.get_decision(judge_id)
     result['choice_a'] = a
     result['choice_b'] = b
-    return jsonify(result)
+    if store.save_session(g.curr_session, g.session_name) is None:
+        before_request()
```
Comment (1heart) [inline hello.py]: See line 58

File: judging_sessions.py [hunk 4]
```diff
@@ -27,16 +27,15 @@ def __init__(self, num_alts):
...
             self.votes[i] = 0
             self.num_times_judged[i] = 0
-            self.q.put((0, i))
 
     def get_decision(self, judge_id):
         if judge_id in self.curr_judges:
             return self.curr_judges[judge_id]
-        a,b  = self.q.get()[1], self.q.get()[1]
+        sorted_x = sorted(self.num_times_judged.items(), key=operator.itemgetter(1), reverse=True)
+        a,b  = sorted_x[0][0], sorted_x[1][0]
```
Comment (1heart) [inline judging_sessions.py:39]: Maybe some out of bounds protection here

File: hello.py [hunk 5]
```diff
@@ -34,50 +31,59 @@ def init_judge_session():
...
 
-@app.route('/results')
+@app.route('/results', methods = ['POST'])
 def results():
     result = dict()
-    if curr_session is None:
-        result['error'] = 'Need to init first!'
-        return jsonify(result)
-    result.update(curr_session.get_results())
+    json_args = request.get_json()
```
Comment (1heart) [inline hello.py]: Remove

File: redis_store.py (hunk 1 above)
Comment (ownzandy) [inline redis_store.py:25]: seems very difficult with redis, so I'm just going to leave it - WATCH + anything doesn't seem to execute anything

**Pull request merged**